client/network/src/utils/json_file.rs

```diff
@@ -0,0 +1,67 @@
+use serde::{de::DeserializeOwned, Serialize};
+use std::{io::Error as IoError, path::Path};
+use tokio::io::AsyncWriteExt;
+
+pub(crate) async fn save(
+	target_file: impl AsRef<Path>,
+	data: &impl Serialize,
+) -> Result<(), IoError> {
+	let target_file = target_file.as_ref();
+	let tmp_file = target_file.with_extension(EXTENSION_TMP);
+	let mut tmp_file_rm_guard = MaybeRmOnDrop::new(&tmp_file);
+
+	let data = serde_json::to_vec_pretty(&data)?;
+
+	let mut temp_fd = tokio::fs::OpenOptions::new()
+		.create(true)
+		.truncate(true)
+		.write(true)
+		.open(&tmp_file)
+		.await?;
+
+	temp_fd.write_all(&data).await?;
+	temp_fd.flush().await?;
+	std::mem::drop(temp_fd);
+
+	tokio::fs::rename(&tmp_file, target_file).await?;
+	tmp_file_rm_guard.disarm();
+
+	Ok(())
```

Comment:
You don't really need the extra `MaybeRmOnDrop` type here. You can use the `defer!` macro from `sp_core`. Something like this:

```rust
let rm_on_drop = Some(&tmp_file);
defer! { rm_on_drop.map(std::fs::remove_file) }
// ...
rm_on_drop.take();
Ok(())
```

Reply:
The closure baked into the `DeferGuard` produced by the macro refers the `rm_on_drop`, so it won't be mutably accessible until the guard is dropped.

I wouldn't want to resort to a help of ref-cell here.

Or did I miss something?

Reply:
Yes, you're right. Sorry, I had a brainfart while writing that snippet. But you don't have to use a `RefCell` here; a `Cell` (which is zero-cost and can't panic) will do since the type is `Copy`. It'd actually look something like this:

```rust
let rm_on_drop = Cell::new(Some(&tmp_file));
defer! { rm_on_drop.take().map(std::fs::remove_file) }
// ...
rm_on_drop.take();
Ok(())
```

Reply:
I took liberty to do [that](https://github.com/paritytech/substrate/pull/12051/commits/3738a1201e5d82dca9d2abda393461f3807ce9f9) .

I will totally understand if that change would be banned as completely unrelated to the main subject of the PR though.

Reply:
Yeah, this is actually a decent idea, although modifying that macro's existing behavior when it is already used in the codebase is not a good idea, as you've noticed. (: But making another variant like you did is fine by me. :+1: